### PR TITLE
Update broadcasting transforms to use `bind_new_parameters`

### PR DESCRIPTION
**Pull request by eddddddy**
**Context:**
Part of the effort towards operator immutability. In the end we want to remove `tape.set_parameters` altogether.

**Description of the Change:**
Update `batch_input`, `batch_params`, and `broadcast_expand` to use `bind_new_parameters` instead of `QuantumScript.set_parameters`. Note that `bind_new_parameters` is only called on an operator if it has broadcasted parameters; otherwise the original operator is passed by reference to the new tape.

**Benefits:**
No longer calls `QuantumScript.set_parameters`. 

**Possible Drawbacks:**
Code is slightly more complex, since `tape.get_parameters` returns the parameters without any reference to the corresponding operator, so care must be taken to map the operators to the corresponding tape parameters.

### CODE REVIEW

File: pennylane/transforms/broadcast_expand.py [hunk 1]
```diff
@@ -90,24 +90,76 @@ def broadcast_expand(tape):
...
+            # no batching in the operator; don't copy
+            for b in range(num_tapes):
+                new_preps[b].append(prep)
+
+    ops = [op for op in tape.operations if op not in tape.prep]
+    for op in ops:
+        # determine if any parameters of the operator are batched
+        if op.batch_size and any(
+            qml.math.ndim(p) != op.ndim_params[j] for j, p in enumerate(op.data)
+        ):
```
Comment (albi3ro) [inline pennylane/transforms/broadcast_expand.py:125]: So `op.batch_size` is not sufficient to determine if the operator is batched?

Comment (eddddddy) [reply pennylane/transforms/broadcast_expand.py:125]: Oh wait, yes it is 😅.

File: pennylane/transforms/broadcast_expand.py [hunk 2]
```diff
@@ -90,24 +90,50 @@ def broadcast_expand(tape):
...
+                new_params = tuple(
+                    p if qml.math.ndim(p) == prep.ndim_params[j] else p[b]
+                    for j, p in enumerate(prep.data)
+                )
+                new_prep = qml.ops.functions.bind_new_parameters(prep, new_params)
+                new_preps[b].append(new_prep)
+        else:
+            # no batching in the operator; don't copy
+            for b in range(num_tapes):
+                new_preps[b].append(prep)
```
Comment (albi3ro) [inline pennylane/transforms/broadcast_expand.py:115]: This block may be worth extracting into a private helper function. Something like:
```
def _split_operations(obj_list: List[Operator], num_tapes: int) -> List[Operator]:
```
This can then just be used twice.

**Review by timmysilv — COMMENTED** [review]
we've been kinda avoiding the discussion on the `_prep`, `_ops`, and `operations` properties of QuantumScript. I think I'm ok with making `_prep` public, but I also wonder if we should make `_ops` public to avoid the list comprehension you do here.

overall lgtm, i'll look it over once more next week after discussing ^ that but I see no problems here

**Comment by eddddddy** [comment]
I'm also OK with making `_ops` accessible, but maybe as a class property with a different name to avoid confusion with `tape.operations`. How about `tape.gates`?

File: pennylane/transforms/batch_params.py [hunk 3]
```diff
@@ -37,6 +37,37 @@ def _nested_stack(res):
...
+    at the given indices split.
+    """
+    # for some reason pylint thinks "qml.ops" is a set
+    # pylint: disable=no-member
+    new_ops = [[] for _ in range(num_tapes)]
+    idx = start_index
+
+    for op in ops:
+        # determine if any parameters of the operator are batched
+        if any(i in indices for i in range(idx, idx + len(op.data))):
```
[the remark this answers is not in the document]
Comment (eddddddy) [reply pennylane/transforms/batch_params.py:53]: No, because we only want to unbatch the parameters at the given indices.

File: pennylane/tape/qscript.py [hunk 4]
```diff
@@ -295,6 +295,26 @@ def circuit(self):
...
+    @property
+    def gates(self) -> List[Operator]:
+        """
+        Returns the operations (excluding state preparations) on the quantum script.
+
+        Returns:
+            list[.Operator]: quantum operations
+        """
+        return self._ops
+
```
Comment (albi3ro) [inline pennylane/tape/qscript.py:317]: I think I'm good with adding these as public properties, but we should break these changes out into a separate PR so we can focus on them.

**Review by timmysilv — APPROVED** [review]
🎉

File: pennylane/transforms/batch_params.py [hunk 5]
```diff
@@ -37,6 +37,37 @@ def _nested_stack(res):
     return tuple(stacked_results)
 
 
+def _split_operations(ops, params, indices, start_index, num_tapes):
```
Comment (albi3ro) [inline pennylane/transforms/batch_params.py:40]: I'm a little confused about the difference between `indices` and `start_index`.

Comment (eddddddy) [reply pennylane/transforms/batch_params.py:40]: `indices` corresponds to which parameters of the tape we want to unbatch, where each entry `i` of `indices` corresponds to accessing the tape parameter at `tape._par_info[i]`.

`start_index` represents the above index for the first parameter of the first operation in `ops`. This is necessary if a tape has a prep, in which case the first few parameters in `params` corresponds to the prep and not the non-prep operations. This is only because we need to call `_split_operations` twice: once for the preps and once for the non-prep operations.

Comment (eddddddy) [reply pennylane/transforms/batch_params.py:40]: Maybe `indices` -> `split_indices` and `start_index` -> `first_op_index`? I can definitely clarify this in the docstring as well.

**Review by mudit2812 — APPROVED** [review]
🚀

**Review by albi3ro — APPROVED** [review]

**Comment by eddddddy** [comment]
[sc-38798]
